crates/testing-framework/src/transform/parser.rs

```diff
@@ -0,0 +1,488 @@
+use nom::multi::{many1_count, separated_list0};
+use nom::sequence::{delimited, pair, preceded, separated_pair, terminated};
+use nom::IResult;
+use nom_locate::LocatedSpan;
+
+use std::str::FromStr;
+
+type Input<'inp> = LocatedSpan<&'inp str>;
+type ParseError<'inp> = VerboseError<Input<'inp>>;
+
+impl FromStr for Sexp {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use nom::combinator::all_consuming;
+
+        let span = nom_locate::LocatedSpan::new(s);
+        cut(all_consuming(delim_ws(parse_sexp)))(span)
+            .map(|(_, v)| v)
+            .map_err(parse_error_to_message)
+    }
+}
+
+pub(crate) fn parse_error_to_message(e: nom::Err<ParseError>) -> String {
+    let e = match e {
+        nom::Err::Failure(e) => e,
+        _ => panic!("shouldn't happen because of top-level cut"),
+    };
+    let contexts = e
+        .errors
+        .iter()
+        .rev()
+        .filter_map(|(span, kind)| {
+            if let VerboseErrorKind::Context(c) = kind {
+                Some(format!(
+                    "  {}:{}: {}",
+                    span.location_line(),
+                    span.get_utf8_column(),
+                    c
+                ))
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+    if contexts.is_empty() {
+        e.to_string()
+    } else {
+        format!("Failed to parse the script:\n{}", contexts.join("\n"))
+    }
+}
+
+pub(crate) fn parse_sexp(inp: Input<'_>) -> IResult<Input<'_>, Sexp, ParseError<'_>> {
+    alt((
+        parse_sexp_call,
+        parse_sexp_list,
+        parse_sexp_string,
+        parse_sexp_symbol,
+    ))(inp)
+}
+
+fn parse_sexp_list(inp: Input<'_>) -> IResult<Input<'_>, Sexp, ParseError<'_>> {
+    context(
+        "within generic list",
+        preceded(
+            terminated(tag("("), multispace0),
+            cut(terminated(
+                map(separated_list0(multispace1, parse_sexp), Sexp::list),
+                preceded(
+                    multispace0,
+                    context("closing parentheses not found", tag(")")),
+                ),
+            )),
+        ),
+    )(inp)
+}
+
+fn parse_sexp_string(inp: Input<'_>) -> IResult<Input<'_>, Sexp, ParseError<'_>> {
+    // N.B. escape are rejected by AIR parser, but we simply treat backslash
+    // as any other character
+    map(
+        context(
+            "within string",
+            preceded(
+                tag("\""),
+                cut(terminated(
+                    is_not("\""),
+                    context("closing quotes not found", tag("\"")),
+                )),
+            ),
+        ),
+        Sexp::string,
+    )(inp)
+}
+
+fn parse_sexp_symbol(inp: Input<'_>) -> IResult<Input<'_>, Sexp, ParseError<'_>> {
+    map(
+        recognize(many1_count(alt((
+            value((), alphanumeric1),
+            value((), one_of("_.$")),
```

Comment:
Add `#` as well, as `canon` streams have it.

Reply:
Changed comment char to `;` as well.  Technically, single `#` can be used as a comment marker, but `;` is Lispy de facto standard.